### PR TITLE
Fix duplicates in AdjustableAudioComponent's adjustments

### PR DESCRIPTION
**Pull request by MrTheMake**
Resolves https://github.com/ppy/osu/issues/819

Supersedes https://github.com/ppy/osu-framework/pull/826

### CODE REVIEW

File: osu.Framework/Audio/AdjustableAudioComponent.cs [hunk 1]
```diff
@@ -78,34 +78,46 @@ public void AddAdjustmentDependency(AdjustableAudioComponent component)
...
                 case AdjustableProperty.Balance:
+                    if (balanceAdjustments.Contains(adjustBindable)) return;
                     balanceAdjustments.Add(adjustBindable);
                     break;
                 case AdjustableProperty.Frequency:
+                    if (frequencyAdjustments.Contains(adjustBindable)) return;
                     frequencyAdjustments.Add(adjustBindable);
                     break;
                 case AdjustableProperty.Volume:
+                    if (volumeAdjustments.Contains(adjustBindable)) return;
```
Comment (Jorolf) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: use a hashset instead of a list...?

Comment (Jorolf) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: https://msdn.microsoft.com/en-us/library/bb359438(v=vs.110).aspx#Anchor_7
"The HashSet<T> class is based on the model of mathematical sets and provides high-performance set operations similar to accessing the keys of the Dictionary<TKey, TValue> or Hashtable collections. In simple terms, the HashSet<T> class can be thought of as a Dictionary<TKey, TValue> collection without values.

A HashSet<T> collection is not sorted and cannot contain duplicate elements. If order or element duplication is more important than performance for your application, consider using the List<T> class together with the Sort method."

Comment (MrTheMake) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: Used this class now. Didn't know C# provides such a class.

Comment (huoyaoyuan) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: `HashSet` is designed for better `Contains` operation, while `List` is more array-like.

Comment (Jorolf) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: Me neither until I googled it ;)

File: osu.Framework/Audio/AdjustableAudioComponent.cs [hunk 2]
```diff
@@ -101,11 +101,20 @@ public void AddAdjustment(AdjustableProperty type, BindableDouble adjustBindable
...
-            balanceAdjustments.Remove(adjustBindable);
-            frequencyAdjustments.Remove(adjustBindable);
-            volumeAdjustments.Remove(adjustBindable);
+            switch (type)
+            {
+                case AdjustableProperty.Balance:
+                    balanceAdjustments.Remove(adjustBindable);
+                    break;
+                case AdjustableProperty.Frequency:
+                    volumeAdjustments.Remove(adjustBindable);
```
Comment (turbedi) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: `frequencyAdjustments`, not `volumeAdjustments`

File: osu.Framework/Audio/AdjustableAudioComponent.cs [hunk 3]
```diff
@@ -101,11 +101,20 @@ public void AddAdjustment(AdjustableProperty type, BindableDouble adjustBindable
...
+                    break;
+                case AdjustableProperty.Volume:
+                    frequencyAdjustments.Remove(adjustBindable);
```
Comment (turbedi) [inline osu.Framework/Audio/AdjustableAudioComponent.cs]: `volumeAdjustments`, not `frequencyAdjustments`

**Comment by peppy** [comment]
Did you find out why components are being added twice in the first place?

**Comment by MrTheMake** [comment]
As discussed in discord, I've added an exception when attempting to add an already added dependency. Moreover, the only case where this is attempted is also fixed.

**Pull request merged**